### ISSUE TITLE
Link error when compiling with -fkeep-inline-functions

### ISSUE THREAD

**Issue by pr0g** [google/googletest#2328]
Hi there,

I've been working on trying to add code coverage information for a header-only library and recently stumbled across `-fkeep-inline-functions`. In a simple test case, this gave correct coverage information for functions declared and defined in a .h file.

I tried building my application with googletest passing this flag but got several link errors:

```
Undefined symbols for architecture x86_64:
  "testing::Expectation::Expectation()", referenced from:
      testing::Sequence::Sequence() in test-mat.cpp.o
      testing::Sequence::Sequence() in test-mat.cpp.o
      testing::Sequence::Sequence() in test-vec.cpp.o
      testing::Sequence::Sequence() in test-vec.cpp.o
      testing::Sequence::Sequence() in test-math.cpp.o
      testing::Sequence::Sequence() in test-math.cpp.o
  "testing::Expectation::~Expectation()", referenced from:
      std::_Sp_counted_ptr<testing::Expectation*, (__gnu_cxx::_Lock_policy)2>::_M_dispose()        in test-mat.cpp.o
      testing::Expectation::Expectation(testing::internal::ExpectationBase&) in test-mat.cpp.o
      testing::Expectation::Expectation(testing::internal::ExpectationBase&) in test-mat.cpp.o
      std::__shared_count<(__gnu_cxx::_Lock_policy)2>::__shared_count<testing::Expectation*>(testing::Expectation*) in test-mat.cpp.o
      std::__shared_count<(__gnu_cxx::_Lock_policy)2>::__shared_count<testing::Expectation*>(testing::Expectation*) in test-mat.cpp.o
      void __gnu_cxx::new_allocator<std::_Rb_tree_node<testing::Expectation> >::destroy<testing::Expectation>(testing::Expectation*) in test-mat.cpp.o
      testing::ExpectationSet::ExpectationSet(testing::internal::ExpectationBase&) in test-mat.cpp.o
      ...
  "testing::internal::g_gmock_mutex", referenced from:
      testing::internal::ExpectationBase::is_retired() const in test-mat.cpp.o
      testing::internal::ExpectationBase::Retire()    in test-mat.cpp.o
      testing::internal::ExpectationBase::call_count() const in test-mat.cpp.o
      testing::internal::ExpectationBase::IncrementCallCount()     in test-mat.cpp.o
      testing::internal::ExpectationBase::IsOverSaturated() const in test-mat.cpp.o
      testing::internal::ExpectationBase::IsSaturated() const in test-mat.cpp.o
      testing::internal::ExpectationBase::IsSatisfied() const in test-mat.cpp.o
      ...
  "testing::internal::GetFailureReporter()", referenced from:
      testing::internal::Assert(bool, char const*, int, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&) in test-mat.cpp.o
      testing::internal::Expect(bool, char const*, int, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&) in test-mat.cpp.o
      testing::internal::Assert(bool, char const*, int, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&) in test-vec.cpp.o
      testing::internal::Expect(bool, char const*, int, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&) in test-vec.cpp.o
      testing::internal::Assert(bool, char const*, int, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&) in test-math.cpp.o
      testing::internal::Expect(bool, char const*, int, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&) in test-math.cpp.o
```

Should it be possible to compile googletest with this flag? It would make my code coverage numbers a lot more realistic (currently I apparently have 99% coverage which is definitely not the case :P)

I'm building on macOS with g++-7 (not AppleClang)

Thanks for your help!

**Comment by pr0g** [comment]
Quick update on this - I was able to fix the link errors by moving some functions in googletest from the .hh files to the corresponding .cc files - I'll create a fork soon with the edits for visibility and use it for my coverage stuff. There probably is a better fix once I understand what's going on a little better but for now my coverage results are much more accurate! 😄 (I'm in fact at around 60%, not 99% 😭)

**Comment by pr0g** [comment]
FYI here is a link to the fork and commit to fix the issue I was seeing - https://github.com/pr0g/googletest/commit/14ab8ee108afa3b8a6a65ec68240bebb368468f7

All I've done is move a few functions from gmock-spec-builders.h to gmock-spec-builders.cc and gmock-internal-utils.h to gmock-internal-utils.cc

There likely is a better solution but this will unblock me for now - I'd be interested to hear what you think.

Thanks!

**Comment by pr0g** [comment]
Some bad news... I realized there was actually another issue I ~~fixed~~ hacked locally when building my app with googletest which I haven't been able to resolve on the travis build machines. It's to do with `recursive_init_error` in `cxxabi.h` when using `-fkeep-inline-functions`. It, unfortunately, does not have a destructor implementation.

```c++
// /usr/local/Cellar/gcc@7/7.4.0/include/c++/7.4.0/cxxabi.h
class recursive_init_error: public std::exception
{
public:
  recursive_init_error() throw() { }
  virtual ~recursive_init_error() throw ();
};
```
```bash
"vtable for __gnu_cxx::recursive_init_error", referenced from:
      __gnu_cxx::recursive_init_error::recursive_init_error() in main.cpp.o
```

I added one locally and that did fix the problem, but trying to resolve this on travis got pretty hairy, so I've removed `-fkeep-inline-functions` and am not using my googletest fork anymore.

It does seem the only way to get reliable coverage information for header-only libraries is to use that flag (no other flags seem to do what I want 😔). One possible solution I thought was to use `-fkeep-inline-functions` and try and build googletest without exceptions so `recursive_init_error` would not get linked in, but when I tried that I still seemed to get the same link error 😩)

I'm not sure if this is something that comes up often but if anyone has any other suggestions or concrete information on how to get reliable coverage for free functions in header files I would be very grateful to hear! 😄

Thanks!

**Comment by pr0g** [comment]
FYI I gave up attempting to use this flag as I just kept running into too many issues 😔

A workaround I found which I'm pretty happy with is to mark up my API with a `#define` that's an alias for `__attribute__((__used__))` when using GCC and Coverage is enabled.

This is giving me accurate coverage information now, the only downside being I need to remember to add the `API #define` in front of my inlined functions.
